transport-native-io_uring/src/main/java/io/netty/channel/uring/AbstractIOUringServerChannel.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelOutboundBuffer;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.ServerChannel;
+
+import java.net.SocketAddress;
+
+abstract class AbstractIOUringServerChannel extends AbstractIOUringChannel implements ServerChannel {
+
+    AbstractIOUringServerChannel(int fd) {
+        super(null, new LinuxSocket(fd));
+    }
+
+    AbstractIOUringServerChannel(LinuxSocket fd) {
+        super(null, fd);
+    }
+
+    @Override
+    protected AbstractUringUnsafe newUnsafe() {
+        return new UringServerChannelUnsafe();
+    }
+
+    @Override
+    protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+        throw new UnsupportedOperationException();
+    }
+
+    public AbstractIOUringChannel getChannel() {
+        return this;
+    }
+
+    abstract Channel newChildChannel(int fd) throws Exception;
+
+    final class UringServerChannelUnsafe extends AbstractIOUringChannel.AbstractUringUnsafe {
+        private final byte[] acceptedAddress = new byte[26];
+
+        @Override
+        public void connect(final SocketAddress remoteAddress, final SocketAddress localAddress,
+                            final ChannelPromise promise) {
+            promise.setFailure(new UnsupportedOperationException());
+        }
+
+        @Override
+        public void uringEventExecution() {
+            final IOUringEventLoop ioUringEventLoop = (IOUringEventLoop) eventLoop();
+            IOUringSubmissionQueue submissionQueue = ioUringEventLoop.getRingBuffer().getIoUringSubmissionQueue();
+
+            long eventId = ioUringEventLoop.incrementEventIdCounter();
+            final Event event = new Event();
+            event.setId(eventId);
+            event.setOp(EventType.ACCEPT);
+            event.setAbstractIOUringChannel(getChannel());
+
+            //Todo get network addresses
+            submissionQueue.add(eventId, EventType.ACCEPT, getChannel().getSocket().getFd(), 0, 0, 0);
```

Comment:
I don't see a good existing method to get the address in a convenient format, so just add some JNI methods to allocate some direct byte buffers with this data. Then you can use `io.netty.channel.unix.Buffer.memoryAddress()` to get the pointers out that you need here.

Reply:
You can also store all of the relevant data on the `Event` object, and change `submissionQueue.add` to take the event object itself as a parameter.

Then when the event loop needs to re-enqueue the ACCEPT comment, it just needs to set a new event id and send the same event object back. Then you'd get the address cached for free.